### PR TITLE
Inital Checkin for adding a MetricTrait to a Kubernetes Deployment Wo…

### PR DESCRIPTION
**Pull request by ddrechse**
…rkload

# Description

Please include a summary of the change and which issue is fixed.  If there are any dependencies, for example on a PR in another repository, please list those.

This PR adds functionality such that a MetricsTrait can be associated with a standard Kubernetes Deployment Workload. There is a new WorkloadDefinition. Prior to this change, Workload Definitions had a children relationship with the Kinds that made up a Workload.  For the Deployment, this relationship became a peer.

# Checklist 

As the author of this PR, I have:

- [ x] Checked that I included or updated copyright and license notices in all files that I altered
- [ x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [x ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_k8sdeployment.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2020, 2021, Oracle and/or its affiliates.
```
Comment (kminder) [inline platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_k8sdeployment.yaml:1]: A nitpick but these copyright for new files should just be 2021.

File: application-operator/controllers/metricstrait/metricstrait_controller.go [hunk 2]
```diff
@@ -701,6 +701,11 @@ func (r *Reconciler) fetchTraitDefaults(ctx context.Context, workload *unstructu
 		return r.NewTraitDefaultsForGenericWorkload()
 	}
 
+	// Match any version of APIVersion=apps and Kind=Deployment
```
Comment (kminder) [inline application-operator/controllers/metricstrait/metricstrait_controller.go:704]: apps is the Group right?

File: application-operator/controllers/metricstrait/metricstrait_controller_test.go [hunk 3]
```diff
@@ -383,6 +383,154 @@ func TestMetricsTraitCreatedForVerrazzanoWorkload(t *testing.T) {
...
+		})
+	// Expect a call to list the child Deployment resources of the containerized workload definition
+	mock.EXPECT().
+		List(gomock.Any(), gomock.Not(gomock.Nil()), gomock.Any()).
+		DoAndReturn(func(ctx context.Context, list *unstructured.UnstructuredList, opts ...client.ListOption) error {
+			assert.Equal("Deployment", list.GetKind())
+			return appendAsUnstructured(list, testDeployment)
+		})
+	// Expect a call to list the child Service resources of the containerized workload definition
+	/*	mock.EXPECT().
```
Comment (kminder) [inline application-operator/controllers/metricstrait/metricstrait_controller_test.go:482]: Remove commented out code if it really isn't required.

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_k8sdeployment.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+apiVersion: core.oam.dev/v1alpha2
+kind: WorkloadDefinition
+metadata:
+  name: deployments.apps
+spec:
+  childResourceKinds:
+  - apiVersion: apps/v1
+    kind: Deployment
```
Comment (kminder) [inline platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_k8sdeployment.yaml:11]: Probably worth a comment here to help future readers understand why Deployment is a child of itself.

File: application-operator/controllers/navigation/unstructured.go [hunk 5]
```diff
@@ -78,6 +78,11 @@ func FetchUnstructuredChildResourcesByAPIVersionKinds(ctx context.Context, cli c
 			return nil, err
 		}
 		for i, item := range resources.Items {
+			// The Kubernetes Deployment Case where Workload is the Child
```
Comment (desagar) [inline application-operator/controllers/navigation/unstructured.go:81]: a more detailed comment here would be useful. I'm not sure I understand why "workload is the child" in this case

File: Jenkinsfile [hunk 6]
```diff
@@ -455,6 +455,14 @@ pipeline {
                                 }
                             }
                         }
+                        stage('k8s deployment workload metrics') {
```
Comment (desagar) [inline Jenkinsfile:458]: are we planning to have more tests for Deployment besides just metrics? It seems too granular to have an entire stage and test suite just for metrics.

Comment (ddrechse) [reply Jenkinsfile:458]: idk, not currently. If a future need arises we can just move stuff around

File: application-operator/controllers/navigation/unstructured.go (hunk 5 above)
Comment (ddrechse) [reply application-operator/controllers/navigation/unstructured.go:81]: this gets into the general design of OAM.  Take for example ContainerizedWorkload. This is an OAM construct that combines a deployment and a service.  Every other OAM workload we have is similar in that it is an OAM construct that refers to children.  Except for native k8s types such as a Deployment, in this case the Workload is the child.  Did that help? ;-)

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_k8sdeployment.yaml (hunk 4 above)
Comment (ddrechse) [reply platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_k8sdeployment.yaml:11]: Done

File: application-operator/controllers/metricstrait/metricstrait_controller_test.go (hunk 3 above)
Comment (ddrechse) [reply application-operator/controllers/metricstrait/metricstrait_controller_test.go:482]: Good catch
Done

File: application-operator/controllers/metricstrait/metricstrait_controller.go (hunk 2 above)
Comment (ddrechse) [reply application-operator/controllers/metricstrait/metricstrait_controller.go:704]: Fixed the others also

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_k8sdeployment.yaml (hunk 1 above)
Comment (ddrechse) [reply platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_k8sdeployment.yaml:1]: right u are
Done

**Review by kminder — APPROVED** [review]
Approved.

**Review by desagar — APPROVED** [review]

**Pull request merged**